README.md
Yaoadk - Yet Another Open Accessory Development Kit
===================================================

This is alternative ADK library package.

We will strive to support the various Arduino IDE and Arduino based ADK boards.

Disclaimer
----------
This software is no warranty.

We are not responsible for any results caused by the use of this software.

Please use the responsibility of the your self.

License
-------
Copyright &copy; 2012 ADK Study Group Tokyo

Licensed under the [Apache License, Version 2.0][Apache]

Original copyright
------------------
+ Adb copyright (c) 2011 Niels Brouwers
+ Android Accessory copyright (c) 2011 The Android Open Source Project
+ USB_Host_Shield copyright (c) 2009-2011 Oleg Mazurov, Circuits At Home


-----------------------------------------------------------------

Yaoadk - ADK/MicroBridgeのためのArduinoライブラリ集
====================================================

Androidとの接続にかかわる様々な 八百万（やおろず）の問題を解決できるライブラリ集 にしていこうと考えています。

詳細な解説については[Yaoadk,-Yet-Another-Open-Accessory-Development-Kit wiki][wiki]を参照ください。

免責事項
--------
本ソフトウェアは無保証です。
ADK勉強会(東京)は本ソフトウェアの利用によって生じた結果に対して一切の責任を負いません。利用者の自己責任のもとご利用ください。

著作権について
--------------
Copyright &copy; 2012 ADK Study Group Tokyo

Licensed under the [Apache License, Version 2.0][Apache]

[Apache]: http://www.apache.org/licenses/LICENSE-2.0
[wiki]: https://github.com/ADKstudyGroupTokyo/Yaoadk/wiki/Yaoadk,-Yet-Another-Open-Accessory-Development-Kit


ChangeLog
2012-08-25  Yuuichi Akagawa <Yuuichi.Akagawa@gmail.com>

        * Fix for Adb USB bug (Cannot connect when configuration descriptor size is over 64 bytes)
        - Adb/usp.cpp

2012-06-25  Yuuichi Akagawa <Yuuichi.Akagawa@gmail.com>
        * Refix "Error: OSCOKIRQ failed to assert" on Mega ADK.  change counter size grow up to 16 bit.
	- USB_Host_Shield/Max3421e.cpp
	- Adb/max3421e.cpp

2012-06-23  Yuuichi Akagawa <Yuuichi.Akagawa@gmail.com>

	* ADK_release_20120606
        - AndroidAccessory/AndroidAccessory.cpp
        - AndroidAccessory/AndroidAccessory.h

        * Fix "Error: OSCOKIRQ failed to assert" on Mega ADK
	- USB_Host_Shield/Max3421e.cpp
	- Adb/max3421e.cpp

2012-06-05  Yuuichi Akagawa <Yuuichi.Akagawa@gmail.com>

	* Arduino Leonardo support.
	- Adb/max3421e.cpp
	- Adb/max3421e.h
	- Adb/max3421e_constants.h
	- USB_Host_Shield/Max3421e.cpp
	- USB_Host_Shield/Max3421e.h
	- USB_Host_Shield/Max3421e_constants.h

2012-05-03  Yuuichi Akagawa <Yuuichi.Akagawa@gmail.com>

	* Fix for the Arduino IDE version compatibility.
	- Adb/examples/Logcat/Logcat.pde
	- Adb/examples/Shell/Shell.pde

2012-04-30  Yuuichi Akagawa <Yuuichi.Akagawa@gmail.com>

	* Fix for the Arduino IDE version compatibility and Mega ADK support.
	- Adb/Adb.h
	- Adb/Usb.cpp
	- Adb/Usb.h
	- Adb/max3421e.cpp
	- Adb/max3421e.h
	- Adb/max3421e_constants.h

2012-03-14  Yuuichi Akagawa <Yuuichi.Akagawa@gmail.com>

	* Unite USB_Host_Shield codes.
	- USB_Host_Shield/ch9.h
	- USB_Host_Shield/Max_LCD.cpp
	- USB_Host_Shield/Max_LCD.h
	- USB_Host_Shield/Max3421e.cpp
	- USB_Host_Shield/Max3421e.h
	- USB_Host_Shield/Max3421e_constants.h
	- USB_Host_Shield/Usb.cpp
	- USB_Host_Shield/Usb.h
	- Remove USB_Host_Shield/circuit directory
	- Remove USB_Host_Shield/demokit directory

	* README.md : New

2012-03-14  Kenichi Yoshida

	* remove CapacitiveSense library

2012-03-08  Kenichi Yoshida

	* First commit


Adb/max3421e_constants.h
/*
  Copyright 2012 ADK Study Group Tokyo

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  Changes: 
    Fix for the Arduino IDE version compatibility.
    Supoort ADK boards
      Google ADK compatible boards(RT-ADK, Arduino MegaADK)
      Duemilanove/UNO + Sparkfun USB Host Shield(DEV-09947)
      Pro mini + USB Host Shield for Arduino Pro Mini

 -----------------------------------------------------------------------------------
 * Copyright 2009-2011 Oleg Mazurov, Circuits At Home, http://www.circuitsathome.com
 * MAX3421E USB host controller support
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the authors nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/* MAX3421E register/bit names and bitmasks */

#ifndef _MAX3421Econstants_h_
#define _MAX3421Econstants_h_

/* SPI pins for diffrent Arduinos */
#define setRSTPIN() pinMode( MAX_RESET, OUTPUT );
#define set_SS(x) digitalWrite(MAX_SS, x)
#define readINT() digitalRead(MAX_INT)
#define readGPX() (LOW)
#define setRST(x) digitalWrite(MAX_RESET, x)

// for the ADK reference board
#define ADK_REF_BOARD

#if defined(__AVR_ATmega1280__) || (__AVR_ATmega2560__)
  #define SCK_PIN   52
  #define MISO_PIN  50
  #define MOSI_PIN  51
  #define SS_PIN    53

  #define MAX_SS    SS_PIN

  #ifdef ADK_REF_BOARD
    #define MAX_INT 9
    #define MAX_GPX 8
    #define MAX_RESET 7

    #define setRSTPIN() (DDRJ |= 0x04)
    #define set_SS(x) digitalWrite(MAX_SS, x)
    #define readINT() ((PORTE & 0x40) >> 6)
    #define setRST(x) { if (x) PORTJ |= 0x04; else PORTJ &= ~0x04; }
  #else
    #define MAX_INT   48
    #define MAX_GPX   49
    #define MAX_RESET 46
  #endif
#endif
#if  defined(__AVR_ATmega168__) || defined(__AVR_ATmega328P__)
  #define SCK_PIN   13
  #define MISO_PIN  12
  #define MOSI_PIN  11
  #define SS_PIN    10

  #define MAX_SS    SS_PIN
  #define MAX_INT   9
  #define MAX_GPX   8
  #define MAX_RESET 7
#endif

/* "Breakpoint" pins for debugging */
//#define BPNT_0      3
//#define BPNT_1      2
     
//#define Select_MAX3421E     digitalWrite(MAX_SS,LOW)
//#define Deselect_MAX3421E   digitalWrite(MAX_SS,HIGH)

/* */

#define ON  true
#define OFF false

/* VBUS states */
#define SE0     0
#define SE1     1
#define FSHOST  2
#define LSHOST  3

/* MAX3421E command byte format: rrrrr0wa where 'r' is register number  */
//
// MAX3421E Registers in HOST mode. 
//
#define rRCVFIFO    0x08    //1<<3
#define rSNDFIFO    0x10    //2<<3
#define rSUDFIFO    0x20    //4<<3
#define rRCVBC      0x30    //6<<3
#define rSNDBC      0x38    //7<<3

#define rUSBIRQ     0x68    //13<<3
/* USBIRQ Bits  */
#define bmVBUSIRQ   0x40    //b6
#define bmNOVBUSIRQ 0x20    //b5
#define bmOSCOKIRQ  0x01    //b0

#define rUSBIEN     0x70    //14<<3
/* USBIEN Bits  */
#define bmVBUSIE    0x40    //b6
#define bmNOVBUSIE  0x20    //b5
#define bmOSCOKIE   0x01    //b0

#define rUSBCTL     0x78    //15<<3
/* USBCTL Bits  */
#define bmCHIPRES   0x20    //b5
#define bmPWRDOWN   0x10    //b4

#define rCPUCTL     0x80    //16<<3
/* CPUCTL Bits  */
#define bmPUSLEWID1 0x80    //b7
#define bmPULSEWID0 0x40    //b6
#define bmIE        0x01    //b0

#define rPINCTL     0x88    //17<<3
/* PINCTL Bits  */
#define bmFDUPSPI   0x10    //b4
#define bmINTLEVEL  0x08    //b3
#define bmPOSINT    0x04    //b2
#define bmGPXB      0x02    //b1
#define bmGPXA      0x01    //b0
// GPX pin selections
#define GPX_OPERATE 0x00
#define GPX_VBDET   0x01
#define GPX_BUSACT  0x02
#define GPX_SOF     0x03

#define rREVISION   0x90    //18<<3

#define rIOPINS1    0xa0    //20<<3

/* IOPINS1 Bits */
#define bmGPOUT0    0x01
#define bmGPOUT1    0x02
#define bmGPOUT2    0x04
#define bmGPOUT3    0x08
#define bmGPIN0     0x10
#define bmGPIN1     0x20
#define bmGPIN2     0x40
#define bmGPIN3     0x80

#define rIOPINS2    0xa8    //21<<3
/* IOPINS2 Bits */
#define bmGPOUT4    0x01
#define bmGPOUT5    0x02
#define bmGPOUT6    0x04
#define bmGPOUT7    0x08
#define bmGPIN4     0x10
#define bmGPIN5     0x20
#define bmGPIN6     0x40
#define bmGPIN7     0x80

#define rGPINIRQ    0xb0    //22<<3
/* GPINIRQ Bits */
#define bmGPINIRQ0 0x01
#define bmGPINIRQ1 0x02
#define bmGPINIRQ2 0x04
#define bmGPINIRQ3 0x08
#define bmGPINIRQ4 0x10
#define bmGPINIRQ5 0x20
#define bmGPINIRQ6 0x40
#define bmGPINIRQ7 0x80

#define rGPINIEN    0xb8    //23<<3
/* GPINIEN Bits */
#define bmGPINIEN0 0x01
#define bmGPINIEN1 0x02
#define bmGPINIEN2 0x04
#define bmGPINIEN3 0x08
#define bmGPINIEN4 0x10
#define bmGPINIEN5 0x20
#define bmGPINIEN6 0x40
#define bmGPINIEN7 0x80

#define rGPINPOL    0xc0    //24<<3
/* GPINPOL Bits */
#define bmGPINPOL0 0x01
#define bmGPINPOL1 0x02
#define bmGPINPOL2 0x04
#define bmGPINPOL3 0x08
#define bmGPINPOL4 0x10
#define bmGPINPOL5 0x20
#define bmGPINPOL6 0x40
#define bmGPINPOL7 0x80

#define rHIRQ       0xc8    //25<<3
/* HIRQ Bits */
#define bmBUSEVENTIRQ   0x01   // indicates BUS Reset Done or BUS Resume     
#define bmRWUIRQ        0x02
#define bmRCVDAVIRQ     0x04
#define bmSNDBAVIRQ     0x08
#define bmSUSDNIRQ      0x10
#define bmCONDETIRQ     0x20
#define bmFRAMEIRQ      0x40
#define bmHXFRDNIRQ     0x80

#define rHIEN       0xd0    //26<<3
/* HIEN Bits */
#define bmBUSEVENTIE    0x01
#define bmRWUIE         0x02
#define bmRCVDAVIE      0x04
#define bmSNDBAVIE      0x08
#define bmSUSDNIE       0x10
#define bmCONDETIE      0x20
#define bmFRAMEIE       0x40
#define bmHXFRDNIE      0x80

#define rMODE       0xd8    //27<<3
/* MODE Bits */
#define bmHOST          0x01
#define bmLOWSPEED      0x02
#define bmHUBPRE        0x04
#define bmSOFKAENAB     0x08
#define bmSEPIRQ        0x10
#define bmDELAYISO      0x20
#define bmDMPULLDN      0x40
#define bmDPPULLDN      0x80

#define rPERADDR    0xe0    //28<<3

#define rHCTL       0xe8    //29<<3
/* HCTL Bits */
#define bmBUSRST        0x01
#define bmFRMRST        0x02
#define bmSAMPLEBUS     0x04
#define bmSIGRSM        0x08
#define bmRCVTOG0       0x10
#define bmRCVTOG1       0x20
#define bmSNDTOG0       0x40
#define bmSNDTOG1       0x80

#define rHXFR       0xf0    //30<<3
/* Host transfer token values for writing the HXFR register (R30)   */
/* OR this bit field with the endpoint number in bits 3:0               */
#define tokSETUP  0x10  // HS=0, ISO=0, OUTNIN=0, SETUP=1
#define tokIN     0x00  // HS=0, ISO=0, OUTNIN=0, SETUP=0
#define tokOUT    0x20  // HS=0, ISO=0, OUTNIN=1, SETUP=0
#define tokINHS   0x80  // HS=1, ISO=0, OUTNIN=0, SETUP=0
#define tokOUTHS  0xA0  // HS=1, ISO=0, OUTNIN=1, SETUP=0 
#define tokISOIN  0x40  // HS=0, ISO=1, OUTNIN=0, SETUP=0
#define tokISOOUT 0x60  // HS=0, ISO=1, OUTNIN=1, SETUP=0

#define rHRSL       0xf8    //31<<3
/* HRSL Bits */
#define bmRCVTOGRD  0x10
#define bmSNDTOGRD  0x20
#define bmKSTATUS   0x40
#define bmJSTATUS   0x80
#define bmSE0       0x00    //SE0 - disconnect state
#define bmSE1       0xc0    //SE1 - illegal state       
/* Host error result codes, the 4 LSB's in the HRSL register */
#define hrSUCCESS   0x00
#define hrBUSY      0x01
#define hrBADREQ    0x02
#define hrUNDEF     0x03
#define hrNAK       0x04
#define hrSTALL     0x05
#define hrTOGERR    0x06
#define hrWRONGPID  0x07
#define hrBADBC     0x08
#define hrPIDERR    0x09
#define hrPKTERR    0x0A
#define hrCRCERR    0x0B
#define hrKERR      0x0C
#define hrJERR      0x0D
#define hrTIMEOUT   0x0E
#define hrBABBLE    0x0F

#define MODE_FS_HOST    (bmDPPULLDN|bmDMPULLDN|bmHOST|bmSOFKAENAB)
#define MODE_LS_HOST    (bmDPPULLDN|bmDMPULLDN|bmHOST|bmLOWSPEED|bmSOFKAENAB)


#endif //_MAX3421Econstants_h_
